File: built-in plugins dsp/mb-mic-isolation.cpp
/**
 * MB Mic Isolator
 * Category : effect
 * Type     : microphone
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Background noise isolation
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MIC_ISOLATION_H
#define MB_MIC_ISOLATION_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMicIsolation : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mic-isolation";
    static constexpr const char* PLUGIN_NAME    = "MB Mic Isolator";
    static constexpr const char* PLUGIN_TYPE    = "microphone";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = -40f;  // range [-80, -10]
    float reduction = 20f;  // range [0, 60]
    float attack = 5f;  // range [0.1, 50]
    };

    MbMicIsolation() = default;
    ~MbMicIsolation() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, -80f, -10f);
        params.reduction = std::clamp(params.reduction, 0f, 60f);
        params.attack = std::clamp(params.attack, 0.1f, 50f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Mic Isolator
        return input;
    }
};

#endif // MB_MIC_ISOLATION_H

File: maxbooster7.5/built-in plugins dsp/mb-mic-isolation.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbMicIsolationPlugin: PluginDefinition = {
  id: "mb-mic-isolation",
  slug: "mb-mic-isolation",
  name: "MB Mic Isolator",
  category: "effect",
  type: "microphone",
  version: "1.0.0",
  description: "Background noise isolation",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "threshold",
      name: "Threshold",
      type: "float",
      defaultValue: -40,
      minValue: -80,
      maxValue: -10,
      automatable: true,
    },
    {
      id: "reduction",
      name: "Reduction",
      type: "float",
      defaultValue: 20,
      minValue: 0,
      maxValue: 60,
      automatable: true,
    },
    {
      id: "attack",
      name: "Attack",
      type: "float",
      defaultValue: 5,
      minValue: 0.1,
      maxValue: 50,
      automatable: true,
    },
  ],
  defaultPreset: { threshold: -40, reduction: 20, attack: 5 },
};

export default MbMicIsolationPlugin;
